### PR TITLE
All shops checkboxes for combination edit/bulk-edit forms

### PR DESCRIPTION
**Pull request by zuk3975**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Integrate "Apply to all shops" checkboxes for combination edit/bulk-edit forms. Detailed some remaining issues in "how to test section" bellow the table. Screenshot from 2022-12-16 12-59-14
| Type?             | refacto
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | 
| Related PRs       | 
| How to test?      | Check bellow.
| Possible impacts? | 

**How to test**
Bulk combination form and edit combination form should now have checkboxes when clicked on fields that are multishop. Checkboxes saying "apply to all shops" which when checked, will apply the value to all the shops that combination is associated to. There are some **issues** that should be solved in dedicated PR:
1. `Disabling switch doesn't trigger the "all shops checkbox" to appear`, but it should. This part should be handled in dedicated PR. (for example when you want to disable "low stock threshold" for all combinations you cannot do it now because of this issue. Same goes for bulk action, if you will just enable the field and not click on related input, the checkbox won't appear). ping @jolelievre this is the bug we've discussed during one daily meetup. :arrow_up: 
2. `The combinations list is visually broken` since this PR https://github.com/PrestaShop/PrestaShop/pull/27719. It needs some tweaking when adding multishop checkboxes to the list in a dedicated PR. https://github.com/PrestaShop/PrestaShop/pull/27719 looks like this (its not related to my changes):
Screenshot from 2022-12-16 17-32-46

### CODE REVIEW

File: src/Core/Form/IdentifiableObject/DataFormatter/AbstractFormDataFormatter.php [hunk 1]
```diff
@@ -57,4 +68,42 @@ protected function formatByPath(array $formData, array $pathAssociations): array
 
         return $formattedData;
     }
+
+    protected function formatMultiShopAssociation(string $originalField, string $formattedField): array
```
Comment (zuk3975) [inline src/Core/Form/IdentifiableObject/DataFormatter/AbstractFormDataFormatter.php]: @jolelievre could you check at this?
I wonder if its a good idea to reduce the manual mapping? Also bellow to method `prefixWithModifyAllShops`, I wonder if I could improve it and also if I should extract it to some "string util" component which could be unit tested.

Comment (jolelievre) [reply src/Core/Form/IdentifiableObject/DataFormatter/AbstractFormDataFormatter.php]: Hmm that's a very good idea to avoid adding mapping for all multishop checkboxes, but you can go even further I think this could be automatically handled in the `formatByPath`

Since `formatByPath` checks that the initial data exist before it sets the new formatted value, you can do the same thing for each field with multishop prefix. Thus the multishop field is only added when the initial one is detected. Because that's one of the feature of this formatted it only returns present data (your current implementation seems to add the multishop fields all the time without checking their presence)

This way you don't even need this extra protected method `formatMultiShopAssociation` and, even better, you don't need to split the multishop fields from the regular ones

Comment (jolelievre) [reply src/Core/Form/IdentifiableObject/DataFormatter/AbstractFormDataFormatter.php]: About the string util I guess you mean that the `prefixWithModifyAllShops` is the same logic as in the `CommandBuilder` then indeed it would be interesting to extract this common logic in an util static method for example that can be unit tested. So it should also be used in the command builder then.

Regarding the implementation itself, it will not be the same anymore since yours is based on pure string whereas the initial one was based on PropertyPath that you don't have here. So there are two possible implementations for this class util:
- based on string path
- based on property path

I liked the property path implementation but honestly, it's a detail so I don't mind if you use either of these as long as it's tested and it works none is better than the other. Maybe the string approach is actually more straightforward and reusable.

TLDR; you would need something like `ModifyAllShopsUtil::prefix(string $propertyPath, string $prefix): string`

File: src/PrestaShopBundle/Resources/config/services/core/form/data_formatter.yml [hunk 2]
```diff
@@ -1,7 +1,12 @@
...
 
-  PrestaShop\PrestaShop\Core\Form\IdentifiableObject\DataFormatter\BulkCombinationFormDataFormatter: ~
+  PrestaShop\PrestaShop\Core\Form\IdentifiableObject\DataFormatter\AbstractFormDataFormatter:
+    abstract: true
+    arguments:
+      - !php/const PrestaShopBundle\Form\Admin\Extension\ModifyAllShopsExtension::MODIFY_ALL_SHOPS_PREFIX
 
-  PrestaShop\PrestaShop\Core\Form\IdentifiableObject\DataFormatter\CombinationListFormDataFormatter: ~
+  PrestaShop\PrestaShop\Core\Form\IdentifiableObject\DataFormatter\BulkCombinationFormDataFormatter:
+    parent: PrestaShop\PrestaShop\Core\Form\IdentifiableObject\DataFormatter\AbstractFormDataFormatter
```
Comment (FabienPapet) [inline src/PrestaShopBundle/Resources/config/services/core/form/data_formatter.yml]: I think there is no need to parent on those services, they can inherit from the same abstract but don't need parent services.

For now they don't have any common dependencies, so better disociate them, and keep it simple.

Comment (FabienPapet) [reply src/PrestaShopBundle/Resources/config/services/core/form/data_formatter.yml]: Maybe add a default `bind`  so you don't need parent, and then you can use autowire

Comment (FabienPapet) [reply src/PrestaShopBundle/Resources/config/services/core/form/data_formatter.yml]: https://symfony.com/doc/current/service_container.html#binding-arguments-by-name-or-type

Comment (jolelievre) [reply src/PrestaShopBundle/Resources/config/services/core/form/data_formatter.yml]: Hmm, interesting Besides since we have a dedicated file for formatters we can use the binding safely (no risk that another parameter with similar name uses it unexpectedly)
 So we can do:

```yml
services:
  _defaults:
    bind:
      $modifyAllNamePrefix: !php/const PrestaShopBundle\Form\Admin\Extension\ModifyAllShopsExtension::MODIFY_ALL_SHOPS_PREFIX
```

**Review by SharakPL — DISMISSED** [review]

**Review by lartist — DISMISSED** [review]

**Review by hibatallahAouadni — COMMENTED** [review]
Hello @zuk3975 

Thanks for your PR :rocket: 
I encounter some issues :disappointed: and here's the list:

### First issue:
When I edit the impact price or refence of the product combination and click on checkbox to "Apply changes to all stores", I get Internal server Error 500 in the console and it takes forever loading, see the attached screen record below:

[pr-30590.webm](https://user-images.githubusercontent.com/23308427/210574102-abcbef08-8fa9-4716-b375-ea81f6249d96.webm)

As you can see the new reference has been saved, in another hand the Impact on price is not saved in the shop itself but only in the first shop :exploding_head: 
And it tooks forever!! we need to get back to the Products list and then acces to product page to see the changes.

### Second issue: 
I created 2 groups shop, eachone has 2 shops. Each created shop import data from the default shop. With the Product V1, all shops display the product combinations. But, when enable the Product V2, the shops belong to the second group shop doesn't display the combinations :disappointed: see the attached screen record below:

[multistore_pr-30590.webm](https://user-images.githubusercontent.com/23308427/210575294-9feac0bd-526a-492f-809a-9eb4edefaae3.webm)

**Without you PR**, I encounter a diff problem in this case:

image

It looks like your PR somehow partially fixed this exception :grin:

BTW, I find out that this issue #27497 was fixed when testing your PR (it's fixed even on develop without PR) :grin:
Please check and feedback.

Thanks!

**Comment by jolelievre** [comment]
Hi @hibatallahAouadni 

thanks for your feedback.

**First issue**
The modify for all shops shouldn't be present in the list, we don't handle it in the list for now and I don't think it's planned. If it is it should be handled in a separate PR (we do'nt even have a proper UX for this use case in the list). SO for this PR what should be fixed instead is that this checkbox should NEVER be visible, I think it's only present here as a side effect because the product list uses a `DeltaQuantityType` form type that is common in several forms (product form, combination form, bulk combination form), but in this specific case we need to disable this checkbox via an additional option (@zuk3975 you can add an option on `DeltaQuantityType` to disable this checkbox in `CombinationItemType`).

**Second issue**
It seems like the bug you are mentioning is related to the Shops page, not the product page directly I noticed there were some problems when we create new shops and ask for product duplications It seems like there is a bug when generating in another group (just a question, is the group you created sharing its stock? because it can be related to this feature sometimes).

The bug is not visible in the product page v1 because it doesn't correctly lists the combinations associated to the current shop context, so the bug was already there but not visible because of the old page's instability in multishop. So I think this needs some further investigation to reproduce the problem and this should be fixed in a dedicated issue for the shop page, not the product page.

A good way to check if it comes from the shop page is to first create shops and group shops, and then test a newly created product. Some bugs have been recently fixed by this PR https://github.com/PrestaShop/PrestaShop/pull/30683 including the copy from shop to shop. If everything works fine when the product is handled purely via the product page then it would confirm the bug from shop page.

File: src/Core/Util/String/ModifyAllShopsUtil.php [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+     */
+    public static function prefixFieldPathWithAllShops(string $fieldPath, string $allShopsPrefix): string
+    {
+        /*
+         * Finds all matches between angle brackets.
+         * E.g. for field "[foo][bar]" it will return array of
+         *
+         *    [
+         *        ['[foo]','[bar]'],
+         *        ['foo', 'bar']],
```
Comment (jolelievre) [inline src/Core/Util/String/ModifyAllShopsUtil.php]:
```suggestion
         *        ['foo', 'bar'],
```

File: src/Core/Form/IdentifiableObject/DataFormatter/AbstractFormDataFormatter.php [hunk 4]
```diff
@@ -28,11 +28,23 @@
...
-class AbstractFormDataFormatter
+abstract class AbstractFormDataFormatter
 {
+    /**
+     * @var string
+     */
+    protected $modifyAllNamePrefix;
+
+    public function __construct(
+        string $modifyAllNamePrefix
```
Comment (jolelievre) [inline src/Core/Form/IdentifiableObject/DataFormatter/AbstractFormDataFormatter.php]:
```suggestion
        string $modifyAllNamePrefix = ''
```

This should be optionable not all formatters will need multishop support so the abstract class shouldn't suggest it's mandatory with this kind of constructor

File: src/PrestaShopBundle/Resources/config/services/core/form/data_formatter.yml [hunk 5]
```diff
@@ -1,6 +1,10 @@
 services:
   _defaults:
-    public: true
+    autowire: true
+    bind:
+      $modifyAllNamePrefix: !php/const PrestaShopBundle\Form\Admin\Extension\ModifyAllShopsExtension::MODIFY_ALL_SHOPS_PREFIX
+
+  PrestaShop\PrestaShop\Core\Form\IdentifiableObject\DataFormatter\AbstractFormDataFormatter: ~
```
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/core/form/data_formatter.yml]: I don't think you need this anymore since the construct parameter is already autowired You don't use this service as an abstract/parent service anymore so you should remove it

**Review by jolelievre — COMMENTED** [review]
Thanks @zuk3975 

Just a few suggestions nothing really blocking, but it seems like there are other things to fix in this PR. And I'm not sure I understand all in your description

> Disabling switch doesn't trigger the "all shops checkbox" to appear, but it should. This part should be handled in dedicated PR. (for example when you want to disable "low stock threshold" for all combinations you cannot do it now because of this issue. Same goes for bulk action, if you will just enable the field and not click on related input, the checkbox won't appear). ping @jolelievre this is the bug we've discussed during one daily meetup. 

I tested it and it seems to we working fine, when the field is enabled the checkbox shouldn't be shown right away, it is displayed when you have focus on the input and/or once you have changed its value. From what I tested it seems to match the expected behaviour (it behaves the same in other regular fields that don't need to be enabled and are multi shops)

> The combinations list is visually broken since this PR https://github.com/PrestaShop/PrestaShop/pull/27719. It needs some tweaking when adding multishop checkboxes to the list in a dedicated PR. https://github.com/PrestaShop/PrestaShop/pull/27719 looks like this (its not related to my changes):

This visual glitch could be fixed in this PR, it's been here for a while no and is only visible on multishop I don't think it should be too complex.

However, one thing that needs to be fixed in the list is the presence of the `Modify for all shops` checkbox, it shouldn't be present in this list, it's not designed to be used this way for now. So the input must be configured not to have this checkbox in the context of combination list form item (although I don't think this is related the the visual glitch since it UX bug as present before if I remember correctly).

**Comment by zuk3975** [comment]
@jolelievre I don't think I can simply disable DeltaQuantity because the "modify_all_shops" option is INSIDE it on the inner "delta" input.
About the Disabling switch - we discussed it before. E.g. if you want to disable "low stock threshold alert" there will not appear "allshops" checkbox unless you will also modify the low stock threshold input itself. All of this will also apply to bulk form - e.g. if you want to set impact on price 0, then you could just toggle the disabling switch, but that doesnt trigger the "all shops" checkbox. So you additionally mus focus the input of the impact_on_price  to finally see the checkbox appear.
TLDR - toggling the disabling switch doesn't trigger the "all shops" checkbox.

**Comment by zuk3975** [comment]
@jolelievre Addressed the feedback and fix additional issue found with the stock_threshold (it didn't work before). 
About the Disabling switch - we discussed it before. E.g. if you want to disable "low stock threshold alert" there will not appear "allshops" checkbox unless you will also modify the low stock threshold input itself. All of this will also apply to bulk form - e.g. if you want to set impact on price 0, then you could just toggle the disabling switch, but that doesnt trigger the "all shops" checkbox. So you additionally mus focus the input of the impact_on_price  to finally see the checkbox appear.
TLDR - toggling the disabling switch doesn't trigger the "all shops" checkbox.

@hibatallahAouadni reminder - Removed all_shops checkobox from list so the design looks good again. But I want to remind - this PR is focued on combination form and bulk combination form, it shouldn't impact the combination list edition anymore.

File: src/Core/Form/IdentifiableObject/DataFormatter/BulkCombinationFormDataFormatter.php [hunk 6]
```diff
@@ -42,33 +44,37 @@ class BulkCombinationFormDataFormatter extends AbstractFormDataFormatter
...
             '[price][unit_price]' => '[price_impact][unit_price]',
             '[price][weight]' => '[price_impact][weight]',
-            // Stock data are the trickiest
             '[stock][delta_quantity][delta]' => '[stock][quantities][delta_quantity][delta]',
             '[stock][fixed_quantity]' => '[stock][quantities][fixed_quantity]',
             '[stock][minimal_quantity]' => '[stock][quantities][minimal_quantity]',
             '[stock][stock_location]' => '[stock][options][stock_location]',
             '[stock][low_stock_threshold]' => '[stock][options][low_stock_threshold]',
-            '[stock][low_stock_alert]' => '[stock][options][low_stock_alert]',
+            sprintf(
```
Comment (jolelievre) [inline src/Core/Form/IdentifiableObject/DataFormatter/BulkCombinationFormDataFormatter.php:62]: There's something wrong in the ay we handle `low_stock_threshold` it shouldn't be this complex I think there should be a way to simplify this if we remove the `low_stock_threshold` independent checkbox and rely fully on the  disabling switch feature instead

File: src/PrestaShopBundle/Form/Admin/Type/DeltaQuantityType.php [hunk 7]
```diff
@@ -72,8 +72,10 @@ public function configureOptions(OptionsResolver $resolver)
         $resolver
             ->setDefaults([
                 'delta_label' => $this->trans('Add or subtract items', 'Admin.Global'),
+                'modify_all_shops' => false,
```
Comment (jolelievre) [inline src/PrestaShopBundle/Form/Admin/Type/DeltaQuantityType.php]: The problem with this is that it's the same option that is already used by the extension and that adds automatically the checkboxes. So this specific extra option should have a different name than `modify_all_shops` or else a checkbox mapped to the whole form type could be added

Comment (jolelievre) [reply src/PrestaShopBundle/Form/Admin/Type/DeltaQuantityType.php]:
```suggestion
                'modify_delta_for_all_shops' => false,
```

File: src/PrestaShopBundle/Resources/config/services/core/form/form_data_handler.yml [hunk 8]
```diff
@@ -202,14 +202,14 @@ services:
   PrestaShop\PrestaShop\Core\Form\IdentifiableObject\DataHandler\CombinationFormDataHandler:
     autowire: true
     arguments:
-      $contextShopId: '@=service("prestashop.adapter.legacy.configuration").getInt("PS_SHOP_DEFAULT")'
-      $defaultShopId: "@=service('prestashop.adapter.shop.context').getContextShopID()"
+      $contextShopId: "@=service('prestashop.adapter.shop.context').getContextShopID()"
```
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/core/form/form_data_handler.yml:205]: Nice catch!

File: src/PrestaShopBundle/Form/Admin/Sell/Product/Stock/StockOptionsType.php [hunk 9]
```diff
@@ -94,6 +94,7 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 'default_empty_data' => 0,
                 'empty_view_data' => null,
                 'modify_all_shops' => true,
+                // @todo: need to trigger opening allShopscheckbox on "disabling_switch" change too.
```
Comment (jolelievre) [inline src/PrestaShopBundle/Form/Admin/Sell/Product/Stock/StockOptionsType.php:97]: I still don't understand this bug exactly but I think it's related to the fact that `low_stock_threshold` is split into two fields in the DB, we can check this tomorrow through peer programming

File: src/Core/Form/IdentifiableObject/DataProvider/CombinationFormDataProvider.php [hunk 10]
```diff
@@ -123,7 +124,7 @@ private function extractStockData(CombinationForEditing $combinationForEditing,
             'options' => [
                 'stock_location' => $stockInformation->getLocation(),
                 'low_stock_threshold' => $stockInformation->getLowStockThreshold(),
-                'low_stock_alert' => $stockInformation->isLowStockAlertEnabled(),
+                sprintf('%slow_stock_threshold', DisablingSwitchExtension::FIELD_PREFIX) => $stockInformation->isLowStockAlertEnabled(),
```
Comment (jolelievre) [inline src/Core/Form/IdentifiableObject/DataProvider/CombinationFormDataProvider.php:127]: This one shouldn't be needed The disabling switch is supposed to be initialized based on the value of the associated field So it shouldn't be split into two data, simply indicating that `'low_stock_threshold' => 0,` should automatically disable it

File: src/Core/Form/IdentifiableObject/DataFormatter/BulkCombinationFormDataFormatter.php (hunk 6 above)
Comment (jolelievre) [reply src/Core/Form/IdentifiableObject/DataFormatter/BulkCombinationFormDataFormatter.php:62]: As we discussed this may be worth a refacto for this specific field (including in the command and queries). But it will be done in a dedicated PR

File: src/PrestaShopBundle/Form/Admin/Sell/Product/Stock/StockOptionsType.php (hunk 9 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Form/Admin/Sell/Product/Stock/StockOptionsType.php:97]: As we checked there is a workaround to make this work, as long as you click/focus on the associated field or change its value That's the initial behaviour of the multishop checkbox so I think it's fine for now
We can improve this later if we get feedback that the UX is not good enough

**Review by jolelievre — APPROVED** [review]
Thanks @zuk3975

File: src/PrestaShopBundle/Form/Admin/Sell/Product/Stock/StockOptionsType.php (hunk 9 above)
Comment (zuk3975) [reply src/PrestaShopBundle/Form/Admin/Sell/Product/Stock/StockOptionsType.php:97]: Raised related issue based on our private discussion here https://github.com/PrestaShop/PrestaShop/issues/30750

**Comment by zuk3975** [comment]
Raised a dedicated issue based on our private discussion about the low_stock_threshold here https://github.com/PrestaShop/PrestaShop/issues/30750

**Review by lartist — APPROVED** [review]

**Comment by jolelievre** [comment]
No need for QA this is a partial PR, it will be tested as a whole once the feature is finished
